control-plane/pkg/contract/extensions.go

```diff
@@ -16,7 +16,85 @@
 
 package contract
 
+import v1 "knative.dev/eventing/pkg/apis/eventing/v1"
+
 // IncrementGeneration increments Generation.
 func (x *Contract) IncrementGeneration() {
 	x.Generation++
 }
+
+func NewAllFilter(filters []v1.SubscriptionsAPIFilter) *Filter {
+	return newCollectionFilter(filters, Dialect_ALL)
+}
+
+func NewAnyFilter(filters []v1.SubscriptionsAPIFilter) *Filter {
+	return newCollectionFilter(filters, Dialect_ANY)
+}
+
+func NewNotFilter(f v1.SubscriptionsAPIFilter) *Filter {
+	return &Filter{
+		Dialect: Dialect_NOT,
+		Filters: []*Filter{
+			FromSubscriptionFilter(f),
+		},
+	}
+}
+
+func NewExactFilter(f v1.SubscriptionsAPIFilter) *Filter {
+	return &Filter{
+		Dialect:    Dialect_EXACT,
+		Attributes: f.Exact,
+	}
+}
+
+func NewPrefixFilter(f v1.SubscriptionsAPIFilter) *Filter {
+	return &Filter{
+		Dialect:    Dialect_PREFIX,
+		Attributes: f.Prefix,
+	}
+}
+
+func NewSuffixFilter(f v1.SubscriptionsAPIFilter) *Filter {
+	return &Filter{
+		Dialect:    Dialect_SUFFIX,
+		Attributes: f.Suffix,
+	}
+}
+
+func NewCESQLFilter(f v1.SubscriptionsAPIFilter) *Filter {
+	return &Filter{
+		Dialect:    Dialect_CESQL,
+		Expression: f.SQL,
+	}
+}
+
+func FromSubscriptionFilter(f v1.SubscriptionsAPIFilter) *Filter {
+	switch {
+	case len(f.All) > 0:
+		return NewAllFilter(f.All)
+	case len(f.Any) > 0:
+		return NewAnyFilter(f.Any)
+	case f.Not != nil:
+		return NewNotFilter(*f.Not)
+	case f.Exact != nil:
+		return NewExactFilter(f)
+	case f.Prefix != nil:
+		return NewPrefixFilter(f)
+	case f.Suffix != nil:
+		return NewSuffixFilter(f)
+	case f.SQL != "":
+		return NewCESQLFilter(f)
+	}
+	return nil
+}
```

Comment:
Can we remove the assumption on how the `SubscriptionsAPIFilter` is constructed and its validation?
In particular, I'd like to see us decoupling this logic from the webhook validation to avoid having an implicit dependency between the current limitation of having only one filter.

Of course, only if the semantic of having multiple filters is well defined.

Reply:
I agree it'd be better to decouple that, however it's not clear yet how can we have "multiple dialects" yet still be compliant with the CE Subscription API. 

but I'm fine to move this logic to some other package outside the contract. Is that what you're asking for?

Reply:
I mean that I'd like to not assume that there is only one filter configured in `f` in this translation layer.

This PR seems related to the semantic of having multiple top-level filters: https://github.com/cloudevents/spec/pull/803/files

They are supposed to be treated in `AND`

Reply:
> I mean that I'd like to not assume that there is only one filter configured in f in this translation layer.

I see what you mean now, yeah that totally works for me. We can rely on the webhook's validation.

Fwiw, cloudevents/spec#803 doesn't remove the restriction that each filter has only a single dialect. It introduces a top level "array" of filters, each filter has exactly one dialect. Which is the case here. 
```json
[
  {"prefix": { "type": "knative"} },
  {"sql": "source = 'kafka'"},
]
```